source/common/config/xds_manager_impl.cc

```diff
@@ -0,0 +1,28 @@
+#include "source/common/config/xds_manager_impl.h"
+
+#include "envoy/config/core/v3/config_source.pb.validate.h"
+
+#include "source/common/common/thread.h"
+
+namespace Envoy {
+namespace Config {
+
+absl::Status
+XdsManagerImpl::setAdsConfigSource(const envoy::config::core::v3::ApiConfigSource& config_source) {
+  ASSERT_IS_MAIN_OR_TEST_THREAD();
+  RETURN_IF_NOT_OK(validateAdsConfig(config_source));
+
+  return cm_.replaceAds(config_source);
+}
+
+absl::Status
+XdsManagerImpl::validateAdsConfig(const envoy::config::core::v3::ApiConfigSource& config_source) {
+  auto& validation_visitor = validation_context_.staticValidationVisitor();
+  TRY_ASSERT_MAIN_THREAD { MessageUtil::validate(config_source, validation_visitor); }
```

Comment:
since this is a private method that is called from `setAdsConfigSource` and we already do main/test thread assertion there, do we need a second assert here?

Reply:
This call also starts a try-catch block (see [here](https://github.com/envoyproxy/envoy/blob/40a1dbfea5c064e44939da4c89f783991e800085/source/common/common/thread.h#L327)). This was done because Envoy wants to discourage try-catch blocks other than the control-plane/main-thread path. Eventually, if/when we get rid of exceptions, this try-catch-assert will be removed.

I'm slightly in favor of keeping the assert in the other method so it's the first thing that is validated, and in case the try-catch one is replaced with an exception free code. Let me know if you think otherwise.

Reply:
sgtm